### PR TITLE
Added SegmentTree Data Structure with an example

### PR DESCRIPTION
**Pull request by LashaBuxo**
As mentioned in my created #847 issue, I added a new data structure: Segment Tree 

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTION.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTION.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] Sort by alphabetical order
- [x] I acknowledge that all my contributions will be made under the project's license.

### CODE REVIEW

File: data_structure/segment_trees.cpp [hunk 1]
```diff
@@ -0,0 +1,92 @@
+/* Program describes usage of a popular Data Structure: Segment Trees
+ *  in Finding maximum number and its index on a given Interval
+ *  in logarithmic = Log(N) Time using SegmentTree Data Structures
+ *  with PreProcessing O(N * Log(N)) Time
+ */
+
+#include <assert.h>
```
Comment (Panquesito7) [inline data_structure/segment_trees.cpp:7]: Please use `cassert`.

File: data_structure/segment_trees.cpp [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+#include <math.h>
```
Comment (Panquesito7) [inline data_structure/segment_trees.cpp:8]: Please use `cmath`.

File: data_structure/segment_trees.cpp [hunk 3]
```diff
@@ -0,0 +1,92 @@
+/* Program describes usage of a popular Data Structure: Segment Trees
```
Comment (Panquesito7) [inline data_structure/segment_trees.cpp:1]: Please add copyright notice like in #807.

File: data_structure/segment_trees.cpp [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+#include <assert.h>
+#include <math.h>
+#include <stdio.h>
+
+#include <algorithm>
+#include <iomanip>
+#include <iostream>
+#include <vector>
+
+// Each Node contains it's value and index
```
Comment (Panquesito7) [inline data_structure/segment_trees.cpp:16]: Please use a comment structure like:

```c++
/**
 * Something here
 */
```
**except** inside functions.
Same in other parts of the code.

File: data_structure/segment_trees.cpp [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        BuildTree_Procedural(2 * v, tl, tm);
+        BuildTree_Procedural(2 * v + 1, tm + 1, tr);
+        if (ST.nodes[2 * v].val >= ST.nodes[2 * v + 1].val)
+            ST.nodes[v] = ST.nodes[2 * v];
+        else
+            ST.nodes[v] = ST.nodes[2 * v + 1];
+    }
+}
+
+void BuildTree(std::vector<int> arr) {
```
Comment (Panquesito7) [inline data_structure/segment_trees.cpp:45]: Function not documented.

File: data_structure/segment_trees.cpp [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+    p = getMaximumOnInterval_Procedural(2 * v, tl, tm, l, std::min(r, tm));
+    q = getMaximumOnInterval_Procedural(2 * v + 1, tm + 1, tr,
+                                        std::max(l, tm + 1), r);
+    if (p.val > q.val)
+        return p;
+    else
+        return q;
+}
+
+Node GetMaximumOnInterval(int l, int r) {
```
Comment (Panquesito7) [inline data_structure/segment_trees.cpp:67]: Function not documented.

File: data_structure/segment_trees.cpp [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+    return getMaximumOnInterval_Procedural(1, 1, ST.N, l, r);
+}
+
+int main() {
```
Comment (Panquesito7) [inline data_structure/segment_trees.cpp:71]: Please add a comment structure like:

```c++
/**
 * Main function
 */
```

File: data_structure/segment_trees.cpp [hunk 8]
```diff
@@ -0,0 +1,122 @@
+/**
+ * Copyright 2020 @author Lasha Bukhnikashvili
+ * @file
+ * Program describes usage of a popular Data Structure: Segment Trees
+ * in Finding maximum number and its index on a given Interval
+ * in logarithmic = Log(N) Time using SegmentTree Data Structures
+ * with PreProcessing O(N * Log(N)) Time
+ */
+
+#include <stdio.h>
```
Comment (Panquesito7) [inline data_structure/segment_trees.cpp:10]: Please use the `cstdio` library.

Comment (LashaBuxo) [reply data_structure/segment_trees.cpp]: Done

File: data_structure/segment_trees.cpp [hunk 9]
```diff
@@ -0,0 +1,121 @@
+/**
```
Comment (kvedala) [inline data_structure/segment_trees.cpp:1]: please change the structure to this:
```cpp
/**
  * @file
  * @brief Implementation of [Segment Trees](https://en.wikipedia.org/wiki/Segment_tree)
  * to find maximum number in an array and its index
  * in logarithmic time - \f$O(N * \log(N))\f$.
  * @author [Lasha Bukhnikashvili](author's GitHub profile link here)
  */
```

File: data_structure/segment_trees.cpp [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+#include <vector>
+
+/**
+ * Each Node contains corresponding value
+ * in array and it's index
+ */
+struct Node {
+    int val;
+    int ind;
+} null;
```
Comment (kvedala) [inline data_structure/segment_trees.cpp:25]: 1. using global variables should be avoided as much as possible unless absolutely necessary. Pass variables as function arguments
2. short ambiguous global variable names like `null` will lead to alerts from LGTM checks

File: data_structure/segment_trees.cpp [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+
+/**
+ * Data Structure stores initial array and its nodes,
+ * which has the 4th size of the array = 4*N
+ */
+struct SegmentTree {
+    int N;
+    std::vector<int> arr;
+    std::vector<Node> nodes;
+} ST;
```
Comment (kvedala) [inline data_structure/segment_trees.cpp:35]: another ambiguous global variable

File: data_structure/segment_trees.cpp [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+
+/**
+ * This is a recursive method, which
+ * Builds the Tree in ST struct variable.
+ * It should be called only once for each array.
+ * Works in O(N * log N)
+ */
```
Comment (kvedala) [inline data_structure/segment_trees.cpp:42]: please add description of input function arguments. 
what do `v`, `tl`, `tr` represent? please check the [reference](https://www.doxygen.nl/manual/docblocks.html) and example [code](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/math/complex_numbers.cpp) from the repo that does not use global variables, all functions and variables are documented and organized into class structure and modular in implementation.

File: data_structure/segment_trees.cpp [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+            ST.nodes[v] = ST.nodes[2 * v + 1];
+    }
+}
+
+/**
+ * This method gets initial array and builds
+ * Tree Data Structure based on this array
+ * and stores it in ST struct variable
+ */
+void BuildTree(std::vector<int> arr) {
```
Comment (kvedala) [inline data_structure/segment_trees.cpp:62]: per the Google C++ standards (that we are following for this repo) an optimal way to pass an un-modifiable class variable is as a const reference:
```suggestion
void BuildTree(const std::vector<int> &arr) {
```

**Comment by kvedala** [comment]
Just tried executing and I received `segmentation fault` - this usually happens when the memory access is corrupted.
Screen Shot 2020-06-27 at 9 53 35 PM

File: data_structure/segment_trees.cpp (hunk 9 above)
Comment (kvedala) [reply data_structure/segment_trees.cpp:1]: add your GitHub profile link where it says : * author's GitHub profile link here*

**Comment by LashaBuxo** [comment]
"A better thing would be to rebase your branch to the main master - this will enable the automated checks and tests for your code including formatting it per repo specs.". Sorry about that, I am new in PR's. Should I mark this Upstream/Master branch and make the corresponding commit to it? I don't exactly get what you mean :( 
New Bitmap Image

**Comment by kvedala** [comment]
> "A better thing would be to rebase your branch to the main master - this will enable the automated checks and tests for your code including formatting it per repo specs.". Sorry about that, I am new in PR's. Should I mark this Upstream/Master branch and make the corresponding commit to it? I don't exactly get what you mean :(
> New Bitmap Image

Seems fine now 👍

**Pull request closed without merging**